### PR TITLE
osd/PGLog: should not rollback further than deleted object version

### PR DESCRIPTION
**Pull request by neha-ojha**
When a deleted object becomes a divergent entry in the pg log,
we should not be able to rollback to a version of the deleted
object that doesn't exist.

To avoid this, we need to preserve the original crt of the pg log,
before we update it in rewind_from_head() and use that to decide whether
we can rollback or not in _merge_object_divergent_entries().

Fixes: http://tracker.ceph.com/issues/36739
Signed-off-by: Neha Ojha <nojha@redhat.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by neha-ojha** [comment]
The original bug can be reproduced with the rados:thrash-erasure-code suite using ec-radosbench workload(which performs a lot of deletes).

50 runs of the test:
Before: http://pulpito.ceph.com/nojha-2019-02-07_15:24:50-rados:thrash-erasure-code-wip-36739-2-distro-basic-smithi/
After: http://pulpito.ceph.com/nojha-2019-02-08_05:39:23-rados:thrash-erasure-code-wip-36739-2019-02-07-distro-basic-smithi/ -- most of them are related to bufferlist corruption.

rados: http://pulpito.ceph.com/nojha-2019-02-08_05:52:40-rados-wip-36739-2019-02-07-distro-basic-smithi/
rerun: http://pulpito.ceph.com/nojha-2019-02-08_19:10:02-rados-wip-36739-2019-02-07-distro-basic-smithi/

No related failures in the above runs.

File: src/osd/PGLog.cc [hunk 1]
```diff
@@ -276,6 +276,7 @@ void PGLog::proc_replica_log(
     divergent,
     oinfo,
     olog.get_can_rollback_to(),
+    olog.get_can_rollback_to(),
```
Comment (jdurgin) [inline src/osd/PGLog.cc]: since this is used after rewind_from_head() it looks like we could run into the same issue here

Comment (neha-ojha) [reply src/osd/PGLog.cc]: yeah, I was hoping to fix that in a separate PR.

Comment (jdurgin) [reply src/osd/PGLog.cc]: sounds good

**Comment by neha-ojha** [comment]
@liewegas @jdurgin  I rebased this PR on https://github.com/ceph/ceph/pull/26350 and reran the reproducer.

http://pulpito.ceph.com/nojha-2019-02-11_18:58:45-rados:thrash-erasure-code-wip-test-revert-distro-basic-smithi/

3575095: does not look related, with no backtrace
3575122: looks like https://tracker.ceph.com/issues/36304
3575127: Though the backtrace on this looks similar to http://tracker.ceph.com/issues/36739, the chain of events is different. In the original bug `PG::rewind_divergent_log()` was being called on the backfill target. But here we are hitting the case when `PG::proc_master_log()` calls `merge_log()`, which ends up  calling `rewind_divergent_log()` on the primary. Need to investigate this one further.

File: src/osd/PGLog.h [hunk 2]
```diff
@@ -926,6 +927,8 @@ struct PGLog : DoutPrefixProvider {
     const bool object_not_in_store =
       !missing.is_missing(hoid) &&
       entries.rbegin()->is_delete();
+    ldpp_dout(dpp, 10) << __func__ << ": hoid " << " object_not_in_store: "
```
Comment (dzafman) [inline src/osd/PGLog.h]: I think you meant to have to have this here.

`ldpp_dout(dpp, 10) << __func__ << ": hoid " << hoid << " object_not_in_store: "`

**Comment by neha-ojha** [comment]
https://github.com/ceph/ceph/pull/26996 is an alternate, more complete fix, since reflecting the missing information populated on the backfill target on the primary, seems quite complicated(might require change in the state machine), before the primary goes active. However, the first two commits here still make sense.

**Comment by neha-ojha** [comment]
Closing in favor of https://github.com/ceph/ceph/pull/27200

**Pull request closed without merging**